### ISSUE TITLE
ActiveRecord distinct with sum produces unexpected results

### ISSUE THREAD

**Issue by CodingAnarchy** [rails/rails#33082]
### Steps to reproduce

```
class Foo < ActiveRecord::Base
  (has "price" field)
end

Foo.create(price: 5)
Foo.create(price: 12)
Foo.create(price: 5)

total = Foo.distinct.sum(:price)
```

### Expected behavior

Returned  value should be the sum of all prices, as there are three distinct `foo` records.  (`total == 22`)

### Actual behavior

Returned value is the sum of all `distinct` _prices_. (`total == 17`)

### System configuration
**Rails version**: 5.2.0

**Ruby version**: 2.5.1

**Comment by composerinteralia** [comment]
Related to #16791

**Comment by mohsen-alizadeh** [comment]
@CodingAnarchy 

There is nothing wrong in your example,

When you have three prices *[5,12,5]* actually you have 2 distinct values *[5,12]* and sum of them will be 17 exactly.

Maybe there is some misunderstanding about distinct.

**Comment by kamipo** [comment]
If you expect `Enumerable#sum` to distinct `foo` records, use `Foo.distinct.sum(&:price)` instead.

**Comment by CodingAnarchy** [comment]
I don’t understand why this is being closed here. If I have a `has_many :through` relation, I often need a `distinct` scoping to prevent duplicate records. This should not prevent me from using `sum(:amount)` via the AR database. Summing all the record fields in Ruby is far less performant, for one, and this leads to a lot of hidden errors where the scope is not in the same place as the `sum`.

In Rails 4, this behavior did not arise. `distinct.sum(:amount)` did indeed give the sum of `distinct` applied over all the fields.

Rails 5 has changed this behavior, and for the worse because this result is both unexpected and effectively prevents the use of `sum` on the AR relation just in case that relation changes in the future to have a distinct scoping.

**Comment by kamipo** [comment]
Can you provide actual use case with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

Our test cases expect `Foo.distinct.sum(:price)` returns 17, if changed to return 22, it will break many things.

**Comment by CodingAnarchy** [comment]
It also looks like the "fix" for #16791 is precisely what broke my use case here.

**Comment by arturobr123** [comment]
YEAH @mohsen-alizadeh
Is right, when you use distinct all the duplicate numbers are not taken (in this example, we have 5 12 5 
 you only get 12 5 because distinct take not repeat numbers)

kind regards

**Comment by darwin67** [comment]
@CodingAnarchy 
Could you update the script so that it passes in Rails 4 like you said? I can confirm that it fails in Rails 5 in a very weird way but the reproduction script you provided doesn't pass `v4.2.0` and blows up in `v4.1.0` and `v4.0.0`.

**Comment by CodingAnarchy** [comment]
@darwin67 Here's an updated version of the script.  I tested it locally and it fails in Rails 5, but passes on `v4.2.0` (tested by setting the branch to `4-2-stable` from the rails repository).  Please let me know if that's still not working for you.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails" #, branch: '4-2-stable'
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :charges, force: true do |t|
    t.decimal :amount, precision: 8, scale: 2
  end

  create_table :charges_people, force: true do |t|
    t.integer :charge_id
    t.integer :person_id
  end

  create_table :people, force: true do |t|
    t.integer :location_id
  end

  create_table :locations, force: true do |t|
  end
end

class Location < ActiveRecord::Base
  has_many :people
  has_many :charges, -> { distinct }, through: :people
end

class Person < ActiveRecord::Base
  has_and_belongs_to_many :charges
end

class Charge < ActiveRecord::Base
  has_and_belongs_to_many :people
end

class BugTest < Minitest::Test
  def test_association_stuff
    amt = rand(10) + 10
    second_amt = rand(10) + 25
    total = 2 * amt + second_amt
    charge_a = Charge.create!(amount: amt)
    charge_b = Charge.create!(amount: amt)
    charge_c = Charge.create!(amount: second_amt)
    person_a = Person.create!(charges: [charge_a])
    person_b = Person.create!(charges: [charge_b, charge_c])
    location = Location.create!(people: [person_a, person_b])

    assert_equal(total, location.charges.sum(:amount))
  end
end
```

**Comment by darwin67** [comment]
@CodingAnarchy 
Thank you, I can verify that it works in `v4.2` now.
Just one more thing to clarify, the following should also work since you want to distinct by people right?

```ruby
person_a = Person.create!(charges: [charge_a])
person_b = Person.create!(charges: [charge_a, charge_c])

# => charge_a * 2 + charge_c == location.charges.sum(:amount)
```

**Comment by CodingAnarchy** [comment]
Sorry, I think the script I originally posted had a typo.  The charges should be distinct by ID, not by person.  So it worked like this in Rails 4:

```ruby
person_a = Person.create!(charges: [charge_a])
person_b = Person.create!(charges: [charge_a, charge_c])

# => amt + second_amt == location.charges.sum(:amount) should be true
```

```ruby
person_a = Person.create!(charges: [charge_a, charge_b])
person_b = Person.create!(charges: [charge_a, charge_c])

# => amt * 2 + second_amt == location.charges.sum(:amount) should be true
```

This example is a bit contrived, I admit (it's not what we were actually doing), but it became pretty important when we were doing joins on `has_many` and the like that result in multiples in the result table.

**Comment by darwin67** [comment]
@CodingAnarchy  I see. Then in short, there likely won't be a change in Rails.

Let's take a look at the generated SQL query for `location.charges.sum(:amount)` in both `v4.2.x` and `v5.2.x` (or `master`).

### v4.2.x
```sql
SELECT DISTINCT SUM("charges"."amount") FROM "charges"
INNER JOIN "charges_people" ON "charges"."id" = "charges_people"."charge_id"
INNER JOIN "people" ON "charges_people"."person_id" = "people"."id"
WHERE "people"."location_id" = ?  [["location_id", 2]]
```

### v5.2.x (master)
```sql
SELECT SUM(DISTINCT "charges"."amount") FROM "charges"
INNER JOIN "charges_people" ON "charges"."id" = "charges_people"."charge_id"
INNER JOIN "people" ON "charges_people"."person_id" = "people"."id"
WHERE "people"."location_id" = ?  [["location_id", 2]]
```

As you can see, the difference in the query is the change from `DISTINCT SUM(...)` to `SUM(DISTINCT ...)`.

Doing a `DISTINCT` on `SUM` is actually not very useful and pretty much doesn't do anything. The result of the query in `v4.2.x` is also not what people would expect when combining the two in ActiveRecord, so the change in https://github.com/rails/rails/issues/16791 was actually a fix and the current behavior is the intended and expected one.

Unfortunately, it was a coincidence that it happened to worked for you in `v4.2.x` but it was never intended to be that way. Also another unfortunate news is currently `distinct` doesn't differentiate based on the attribute you pass in and will use `*` by default or whatever the value you passed in with `sum`, therefore the current `arel` doesn't really have a good way to have `distinct` use a specific column while computing the sum in the database.

The only way to get it to work for you now is use the `Enumerable#sum` like @kamipo said.
Which will be something like
```ruby
location.charges.sum(&:amount)
```

**Comment by darwin67** [comment]
> I can only say that the SQL query in 4.2.x did what I expected (at least on Postgres) and did indeed give me the sum of distinct records in my table.

Like I said, you were lucky an unintended behavior worked for you, and it's unfortunate that the correct behavior broke your use case. But no tool is perfect and I would recommend you take a look at the generated SQL queries every once in a while during development and verify it's not generating an uncommon query.

> Furthermore, I would expect that the chaining of scopes would not be affected by sum: i.e., that location.charges.sum(:amount) should be equal to location.charges.sum(&:amount).

In general it shouldn't, but in the case of combining a `COUNT` or `SUM` with `DISTINCT`, it should since there's no way the query builder can tell what you are trying to do with the values.

> My sum can now change in many places because I added a distinct on a join scope somewhere else, which has to be common for has_many :through.

No it isn't (or shouldn't) be a common thing to use `DISTINCT` in the first place even without nesting queries or doing joins. `GROUP BY` is way more efficient and most of the time will give you the same results for what you want with `DISTINCT`.

Let's say this was working like you wanted and ActiveRecord somehow figures that you want your return value to be distinct by id, the query below will likely be what it generates so try running it in your postgres console, 
```sql
SUM(DISTINCT "charges"."id", "charges"."amount") FROM "charges"
INNER JOIN "charges_people" ON "charges"."id" = "charges_people"."charge_id"
INNER JOIN "people" ON "charges_people"."person_id" = "people"."id"
WHERE "people"."location_id" = <location id>
```
and you will see that it happily blows up because it doesn't know what you want to sum it with. You can just pass `"charges"."id"` in and now it will sum up the id values not the amounts.

Hopefully you can see that what you're asking for isn't really intuitive even using raw SQL, it will at least require one level of nesting. (Disclosure: I'm not a SQL expert so someone please correct me if I'm wrong)

> From my perspective, this “fix” makes the SQL sum unpredictable and therefore useless. I am also not satisfied with being told I have to use a more inefficient sum in Ruby or raw SQL - not having to resort to those two options is why I use Rails in the first place.
In my opinion, this fix merely cross-contaminates the different aspects of the problem.

First, I'm sorry it doesn't fully fit your needs, but an abstraction layer will always have that problem of not being able to scratch all of your itchy places. Any reasonable size application will always have some raw SQL or have to utilize the more inefficient solutions somewhere.

Second, Rails is maintained and improved by the public, so you can always submit a pull request for changes if you're not satisfied.


Having said all that, I did came up with something that doesn't require you to use `Enumerable#sum` or raw SQL.
```
Charge.where(id: location.charges.pluck(:id)).sum(:amount)
# => SELECT "charges"."id" FROM "charges" INNER JOIN "charges_people" ON "charges"."id" = "charges_people"."charge_id" INNER JOIN "people" ON "charges_people"."person_id" = "people"."id" WHERE "people"."location_id" = ?  [["location_id", 3]]
# => SELECT SUM("charges"."amount") FROM "charges" WHERE "charges"."id" IN (?, ?, ?)  [["id", 7], ["id", 8], ["id", 9]]
```
It will generate 2 different queries but databases should (hopefully) be smart enough to cache that value so it doesn't end up querying it multiple times.
The reason for doing so should hopefully be obvious by now that what you're doing is not actually common (or I should phrase it as not best practice) even in raw SQL, so you will either have to nest a query or run 2 different ones anyway.

**Comment by CodingAnarchy** [comment]
> you were lucky an unintended behavior worked for you, and it's unfortunate that the correct behavior broke your use case

I think it broke a lot of people's use cases, and the intuitive behavior of the methodology.  In this particular case, if the `DISTINCT SUM([field])` was the wrong behavior, then it should have explicitly been redirected to do a `GROUP BY id` to get the concept across.  Or even automatically implementing a subquery would be preferable.

I can see why the SQL only happened to work in my case, or at least appeared to, and was otherwise broken, but it isn't "fixed" either, in the sense that the intuitive result of `foos.distinct.sum(:bar)` is not the sum of all `bars` on distinct `foos`.

> In general it shouldn't, but in the case of combining a COUNT or SUM with DISTINCT, it should since there's no way the query builder can tell what you are trying to do with the values.

I disagree.  I would find this to be a trivial thing that we should be able to understand.

`foos.distinct.sum(:bar)` should give me the sum of that `bar` field on all the distinct records in the collection.

> No it isn't (or shouldn't) be a common thing to use DISTINCT in the first place even without nesting queries or doing joins.

`has_many :through` can often return multiple join records.  Joins are commonplace, and there are numerous locations in multiple apps that I have used a `distinct` to get singular records for an item.  Particularly common is the `foos.join(:bars).distinct` pattern to get "all foos with attached bars".

See: 

https://stackoverflow.com/questions/5846638/how-to-display-unique-records-from-a-has-many-through-relationship

https://www.lugolabs.com/articles/distinct-active-records-in-has_many-through-associations

> Let's say this was working like you wanted and ActiveRecord somehow figures that you want your return value to be distinct by id, the query below will likely be what it generates

The example you gave is not necessarily the requirement of the query builder.  If the query builder can identify that I have used both distinct and sum, it could build a query that actually gives me the sum of the distinct records.  Even if it does it as a `SELECT SUM(bar) FROM ([previously built DISTINCT query])`.

> First, I'm sorry it doesn't fully fit your needs, but an abstraction layer will always have that problem of not being able to scratch all of your itchy places.

It's not merely that it doesn't fit my needs.  Let me explain in a hypothetical scenario.  Let's say I have some `has_many :bars` association on a `Foo` class.  This works just fine.  Now I want to add a `has_many :bazes, through: :bars`.  Great.  But when I get `foo.bazes`, I could get duplicate records that show up on my use of the collection.  That's not good, but it's easy to solve - I put `-> { distinct }` in the association.  Now it will give me exactly one instance of each `baz`, even if it is associated with multiple `bars`.  But now, suddenly, my DB sums (and counts, and averages, etc.) are giving me completely different answers!

> have to utilize the more inefficient solutions somewhere

I recognize that there are some tradeoffs to be made, but in this case, there is essentially no use case where I would want to combine a `distinct` on some association or join with `sum` in this manner.  I further see no reason why `foos.distinct.sum(:bar)` should not give me the "sum over the distinct records".  It seems further trivial to do a `foos.sum("DISTINCT foos.bar")` if I actually wanted the other for some reason, and no real reason why a nicer syntax could not be added to the query builder for that - something like `foos.sum(distinct: :bar)` or `foos.sum(:bar, distinct: true)`.

> Second, Rails is maintained and improved by the public, so you can always submit a pull request for changes if you're not satisfied.

This would entail a lot of work to remove the change and make it work the way I should, and if I can't even get y'all here to see the problem, it's not something I would deem worth doing just to get it shot down.

> It will generate 2 different queries but databases should (hopefully) be smart enough to cache that value so it doesn't end up querying it multiple times. 

If you did it without pluck, it would just generate a subquery.  But this is besides the point.  I shouldn't need a workaround to avoid using a framework method because that method can suddenly change to do wildly different behavior depending on whether it is chained with another method that might not even be in the same part of the code.

Essentially, as a result of this issue, I am concluding (a) never use `distinct` (use `group_by` instead?), (b) never use the DB `sum`, (c) forget Rails and the DB and do everything in Ruby if at all possible.  And you're telling me that "well, no framework is perfect".  To me, these are serious shortfalls introduced by this change.  

And it's not merely that `sum` is unintuitive, but that it drastically changes behavior based on other code around it.

**Comment by CodingAnarchy** [comment]
Just wanted to note that this bit us again where we had a situation with:

```ruby
class Company
   has_many :customers
   has_many :line_items, -> { distinct }, through: :customers
  ...
end
```
and this in another file (an accounting report service):

```ruby
company.line_items.sum(:amount)
```

As such, we have instituted a company policy to simply not use AR `sum` at all because relationship changes might introduce a `distinct` or `join` somewhere else.  If I ever get a chance, I might need to monkey patch this so I don't lose the performance, but I hope that someone comes across this and understands the problem.

**Comment by CodingAnarchy** [comment]
In the real-world scenario, this was the result:

```
company.line_items.sum(:amount)
=> 4419.9
company.line_items.distinct(false).sum(:amount)
=> 143424.3
company.line_items.sum(&:amount)
=> 56762.3
```

This should not be a "gotcha" as it is now, and I hope the AR team comes around to resolving it to change the SQL generated to combine these more appropriately.

**Comment by caifara** [comment]
Just ran into this issue today while upgrading from Rails 4.2.

For what it's worth, I'm with @CodingAnarchy. I do understand the logic behind the conversation in this thread. I would still like to add my point of view as well: the current way is counter intuitive to the point that even an error would need to be raised. Or … the following reasoning is wrong and I should change several parts of our codebase and try to think about these relations and my life choices :-) in a different way:

* `company.line_items` is supposed to return the line_items of the company. If needed, a distinct is added to the association to make sure the statement returns unique line_items. The distinct part is hidden as an implementation detail and does not leak to other parts of the application to make sure `company.line_items` always returns the intuitive records.
* `company.line_items.sum(:qty)` is supposed to sum the qty's of the line_items that would otherwise be returned by `company.line_items`. Thus `company.line_items.sum(:qty)` and `company.line_items.sum(&:qty)` should return the same amount.
* Even without the hidden distinct call the intuitive meaning of `company.line_items.distinct.sum(:qty)` would be the sum of the qties of the results returned by `company.line_items.distinct`.

If this isn't possible, I'm almost akin to state that an error should be raised.

(Calculating the sum of distinct qties is a more exotic requirement. Intuitively I would suggest: `company.line_items.sum("DISTINCT qty")`.)

@darwin67 would you build this another way?

**Comment by caifara** [comment]
We had some methods like:

```` ruby
def self.total_qty
  sum(:qty)
end
````

To get around this problem (in my view still definitely a bug) we transformed the method like this so we don't need to use pluck:

```` ruby
def self.total_qty
  unscoped.where(id: select(:id)).sum(:qty)
end
````

**Comment by thedarkside** [comment]
I honestly agree with @caifara! Ran into this bug. After upgrading Rails some sums were obviously wrong so i landed here. It feels wrong that a `distinct` call in a nested scope affects the sum value. 
The called `distinct` should ensure to take one row only once into account which is different from the current behaviour where it filters duplicate values.

The suggested solution to use `.sum(&:x)` instead has two drawbacks: 
1. its slower because we are not using the database for calculation. Instead calculate in the Ruby Process.
2. it doesnt work with joined table values

We are now forced to write more code as a workaround. Thats not satisfying at all...

**Comment by joevandyk** [comment]
Just ran into this as well.

**Comment by joevandyk** [comment]
@kamipo thoughts on re-opening this?

**Comment by thebravoman** [comment]
We are also stucked on this. Took us a good morning to debug it and realize sum is totally not working with distinct.

**Comment by trev** [comment]
Ditto.

**Comment by jiggins89** [comment]
still seems to be an issue here, it this being reopened?

**Comment by hathawad** [comment]
We have the same issue going from Rails 4.2 to 5.0

**Comment by tgxworld** [comment]
@kamipo What do you think about supporting something like the following so that we at least have a way to sum in SQL.

`Account.distinct.select(:id, :credit_limit).sum(:credit_limit)` -> `SELECT SUM(credit_limit) FROM (SELECT DISTINCT id, credit_limit FROM accounts) sub_query;`

This will be similar to `count` in some way. 

`Account.distinct.select(:credit_limit).count(:all)` -> `SELECT COUNT(*) FROM (SELECT DISTINCT "accounts"."credit_limit" FROM "accounts") subquery_for_count`

**Comment by kamipo** [comment]
Unless suddenly change the behavior of `Foo.distinct.sum(:price) #=> 17` https://github.com/rails/rails/issues/33082#issue-330333635 (originally it was ambiguous as to which way it should work though), supporting both ways (returning 17 or 22) is fine to me.

**Comment by tgxworld** [comment]
I think it is still ambiguous which way it should work 

What this issue is asking for is really `SELECT SUM(credit_limit) FROM (SELECT DISTINCT id, credit_limit FROM accounts) sub_query;` but what we have now is `SELECT SUM(DISTINCT credit_limit) FROM accounts)`. I figured that if we can support `sum` to build a sub query when `select` is present, we can at least have a way to achieve what this issue is asking for with ActiveRecord.

**Comment by kamipo** [comment]
If a user just does `Account.distinct.select(:id, :credit_limit).sum(:credit_limit)`, I suppose/agree that a user is likely to expect `SELECT SUM(credit_limit) FROM (SELECT DISTINCT id, credit_limit FROM accounts) sub_query`.

But if the case of `relation.distinct.sum(:credit_limit)` (we don't know if a relation has select values or not), a user may expect the current behavior, changing to make the subquery will be considered a breaking change.

**Comment by tgxworld** [comment]
> If a user just does `Account.distinct.select(:id, :credit_limit).sum(:credit_limit)`, I suppose/agree that a user is likely to expect `SELECT SUM(credit_limit) FROM (SELECT DISTINCT id, credit_limit FROM accounts) sub_query`.

Sorry what I was trying to say is that we currently can't do `Account.distinct.select(:id, :credit_limit).sum(:credit_limit)` so I was proposing to support that so that there is at least a way to sum across distinct records through ActiveRecord.

**Comment by kamipo** [comment]
> But if the case of `relation.distinct.sum(:credit_limit)` (we don't know if a relation has select values or not), a user may expect the current behavior, changing to make the subquery will be considered a breaking change.

What I said in the latter sentence is that changing behavior to support that is considered as a breaking change for users who expect the current behavior.

Maybe we need a new way like `sum(:credit_limit, distinct: true)` to avoid the ambiguous and deprecate the current behavior.

**Comment by tgxworld** [comment]
Hmm but in the case that I am proposing, there is an explicit `select` while the current `relation.distinct.sum(:credit_limit)` is ambiguous since we don't know what column the distinct is applied to. The current behavior is that `distinct` is applied to `:credit_limit` but people are actually expecting `:id`. 

I've yet to look how hard this will be to implement but `relation.distinct(:id).sum(:credit_limit)` feels like more intuitive to me.

**Comment by kamipo** [comment]
> The current behavior is that `distinct` is applied to `:credit_limit` but people are actually expecting `:id`.

In this issue report, I see it is expected that `distinct` is applied to `:id`.

But I don't think no one expects `relation.distinct.sum(:credit_limit)` is applied to `:credit_limit` which the behavior that existed for a long time.

**Comment by caifara** [comment]
I feel it's only ambiguous in theory as one expects to be able to read a statement from left to right. Almost all ruby code I can think of adheres to that.

To me, this feels like common sense:

`relation.distinct` => distinct relation by id
`relation.distinct.sum(:credit_limit)` => sum of credit_limit of the rows given by the statement above
`relation.distinct(:credit_limit).sum(:credit_limit)` => sum of distinct credit limits

If you would rethink this from scratch, wouldn't you want this kind of api? Why would someone think the second statement will return the sum of distinct credit_limits? I suppose that would only be the case for people familiar with this issue.

Another solution may be to require the distinct call to have a column name param. That way every call is explicit (and would, just like the examples above, allow to be read from left to right).

**Comment by CodingAnarchy** [comment]
> In this issue report, I see it is expected that `distinct` is applied to `:id`.

I would actually have expected that `relation.distinct.sum(:credit_limit)` would produce something like:

```sql
SELECT SUM(credit_limit) FROM (SELECT DISTINCT * FROM [relation]) as distinct_relation
```

with that then optimized for the particular database you are actually working with.

**Comment by ozzyaaron** [comment]
I think others have already summarized well. I just wanted to add another take and see if it works to convince the maintainers. I also wouldn't want this marked as stale or anything. 

I'm unsure why it is up for debate that calling distinct on the relation is expected to do anything but return a distinct set of that relation and then calling any aggregate calculation on that set would perform the calculation on the distinct set. 

I think the above follows set theory, as well as functional theory (piping results).

To call distinct on the calculated element of the new set to break with reasonable expectations and any math and functional theory I've seen. 

I would expect people that want to actually perform a distinct on the calculation to do something like `relation.sum("DISTINCT element")`

I'm sure others have worked around this themselves already but we have resorted to using `pluck` e.g. `scope.distinct.pluck("SUM(...)").first`

Anyway, as always this is all free so no anger or nothing, it just seems like a strange choice to dig in on keeping the current behaviour. The current behaviour so clearly seems like a bug.

**Comment by oma** [comment]
I'd like to chime in that context matters, in context of has_many the current behavior surprises.

    class Company
      has_many :line_items, -> { distinct_records }, through: : customers
      # distinct_records isn't a thing, just trying to state intent
    end

    #The intent of the following is pretty obvious
    company.line_items.sum(:amount)

So, disregarding the subjective discussion on what is objectively better behavior for distinct, how are we supposed to achieve our goal of `unique records` and sums in that context, that many here are advocating for? 
:pray: :heart:

**Comment by ozzyaaron** [comment]
@oma sorry I'm not sure which people you're pointing at as there seem to be at least two camps of thought here. 

It may be worth updating your comment to include the line of thought that you think is wrong given the example you provided. 

My perspective is that the code you've shown should still sum the `amount` column on the distinct set of `line_item` records. `line_items` is the set (which you disinct-ed some way) and you are calling a sum of amount on that set. 

I'm not so much trying to take a subjective take on this but I'm trying to look at it through the lense of set theory or object oriented programming or functional programming. The current behaviour doesn't make sense from those perspectives so it would be good to hear the other perspective/s. 

Unfortunately the answer seems to be that it is current behaviour. I absolutely get that (and Rails is free after all!) but I, and many others, still think that is a wrong POV to take and the behaviour actually needs to be fixed.

**Comment by oma** [comment]
I'm sorry, I didn't really post to respond to you specifically, I should've been clear about that.

I agree with the piping principle, but I don't see `distinct` changing due to backwards compatibility and since there isn't consensus. 

So, would it be better to define something new rather than changing distinct?
```
has_many_distinct ..?  # noone would not read this as distinct records.
has_many, -> { distinct_records }, ...
```
I would help build, but I'd prefer getting shot down as early as possible :D

**Comment by ozzyaaron** [comment]
No problems @oma  I was honestly just trying to make your comment easier to relate to the discussion as I was a bit unclear on which way you were leaning. Could also just be my reading! 😄 

TLDR; the issue is in ActiveRecord and I think should probably be fixed there as a breaking change for Rails 7.

Just to be clear the problem is not fixed by adding a new association definition method. 

At its core the issue is `Model.distinct.sum(field)` will generate `SUM(DISTINCT field) FROM models` rather than something like `SUM(field) WHERE models.primary_key IN (SELECT models.primary_key FROM models)` or some other form of a SUM over the distinct set. 

You can reproduce the issue (as claimed by myself and others) without using associations at all. 

I would argue that fixing this behaviour could be made a candidate for ActiveRecord 7 as a breaking change in Rails. I get that changing it mid-major is very undesirable but keeping a bug could be just as bad over the longer term. The behaviour has changed from the right behaviour to the wrong behaviour previously (apparently). 

Anyway all just my take and 2c - or maybe even 3c 😄

**Comment by 0llirocks** [comment]
### Environment
ruby: 3.1.2p20
rails: 7.0.7.2

### Steps to reproduce
objects:
{ id: 0, value: 10.0 }, 
{ id: 1. value: 10.0 }

Object.where(id: [0 ,1].distinct.sum(:value)

### Expected outcome
20.0

### Actual outcome
10.0

SQL query: 
```sql
SELECT SUM(DISTINCT "objects"."value") FROM "objects" WHERE "objects"."id" IN ($1, $2)  [["id", 0], ["id", 1]]
```

I didn't expect that the duplicate **value** gets eliminated, took me quite some time to figure it out.

**Comment by gukandrew** [comment]
why? why this issue actual but is already closed?

**Comment by zernie** [comment]
Imo current behaviour needs to be deprecated and proper aggregate methods need to introduced
